src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,235 @@
+        }
+        Ok(())
+    }
+}
+
+// ===== impl InsertTracker =====
+
+impl Clone for Tracker {
+    fn clone(&self) -> Self {
+        self.shared.clone_tracker(self.idx);
+        Self {
+            shared: self.shared.clone(),
+            idx: self.idx,
+            t0: self.t0,
+        }
+    }
+}
+
+impl Drop for Tracker {
+    fn drop(&mut self) {
+        self.shared.drop_tracker(&*self);
+    }
+}
+
+impl Shared {
+    const INITIAL_RECORDERS: usize = 32;
+
+    fn new() -> Self {
+        let mut recorders = Vec::with_capacity(Self::INITIAL_RECORDERS);
+        for i in 0..Self::INITIAL_RECORDERS {
+            let next = AtomicUsize::new(i + 1);
+            recorders.push(Recorder {
+                ref_count: AtomicUsize::new(0),
+                next,
+            })
+        }
+        Self {
+            histogram: Mutex::new(Histogram::default()), // TODO(eliza): should we change the bounds here?
+            recorders: RwLock::new(recorders),
+            next_recorder: AtomicUsize::new(0),
+        }
+    }
+
+    fn tracker(self: Arc<Self>) -> Tracker {
+        let t0 = Instant::now();
+        loop {
+            let idx = self.next_recorder.load(Ordering::Relaxed);
+            // This is determined in a scope so that we can move `Self` into the
+            // new tracker without doing a second (unecessary) arc bump.
+            let acquired = {
+                // Do we have any free recorders remaining, or must we grow the
+                // slab?
+                let recorders = self
+                    .recorders
+                    .read()
+                    .ok()
+                    .filter(|recorders| idx < recorders.len())
+                    .unwrap_or_else(|| {
+                        // If there are no free recorders in the slab, extend it
+                        // (acquiring a write lock temporarily).
+                        self.grow();
+                        self.recorders.read().unwrap()
+                    });
+
+                let next = recorders[idx].next.load(Ordering::Acquire);
+                // If the recorder is still idle, update its ref count & set the
+                // free index to point at the next free recorder.
+                recorders[idx]
+                    .ref_count
+                    .compare_and_swap(0, 1, Ordering::AcqRel)
+                    == 0
+                    && self
+                        .next_recorder
+                        .compare_and_swap(idx, next, Ordering::AcqRel)
+                        == idx
+            };
+
+            if acquired {
+                return Tracker {
+                    shared: self,
+                    idx,
+                    t0,
+                };
+            }
+
+            // The recorder at `idx` was not actually free! Try again with a
+            // fresh index.
+            atomic::spin_loop_hint()
+        }
+    }
+
+    #[cold]
+    #[inline(never)]
```

Comment:
I have an idea of what this attribute does, but what about this function made you add it?

Reply:
I'm also curious.

Reply:
`#[cold]` is a hint to the compiler that it should use a calling convention that improves performance at the callsite under the assumption that a function is called infrequently: https://llvm.org/docs/LangRef.html#calling-conventions

in retrospect, I realise that `#[cold]` implies that a function should not be inlined, so I think the `#[inline(never)]` should be removed.